sensor/kubernetes/listener/resources/selector_test.go

```diff
@@ -0,0 +1,221 @@
+	return ""
+}
+
+func (m mockSelector) Add(r ...labels.Requirement) labels.Selector {
+	return nil
+}
+
+func (m mockSelector) Requirements() (requirements labels.Requirements, selectable bool) {
+	return nil, false
+}
+
+func (m mockSelector) DeepCopySelector() labels.Selector {
+	return nil
+}
+
+func (m mockSelector) RequiresExactMatch(label string) (value string, found bool) {
+	return "", false
+}
+
+func (m mockSelector) Matches(labels labels.Labels) bool {
+	m.testSuite.hasMatchesBeenCalled = true
+	return m.internalSelector.Matches(labels)
+}
+
+func (s *SelectorWrapperTestSuite) injectMockSelector(sw *selectorWrap) {
+	sw.selector = mockSelector{sw.selector, s}
+}
+
+func (s *SelectorWrapperTestSuite) TestLabelMatching() {
+	tests := map[string]struct {
+		givenSelectorLabels                 map[string]string
+		matchEmptySelector                  selectorWrapOption
+		givenMatchingLabels                 map[string]string
+		expectedMatch                       bool
+		expectedSelectorMatchesFunctionCall bool
+	}{
+		"Empty selector with matchEmpty set to false should match nothing; attempting to match some labels": {
+			givenSelectorLabels:                 labelsEmpty,
+			matchEmptySelector:                  emptyMatchesNothing(),
+			givenMatchingLabels:                 labelsThreeElements,
+			expectedMatch:                       false,
+			expectedSelectorMatchesFunctionCall: false,
+		},
+		"Empty selector with matchEmpty set to false should match nothing; attempting to match empty labels": {
+			givenSelectorLabels:                 labelsEmpty,
+			matchEmptySelector:                  emptyMatchesNothing(),
+			givenMatchingLabels:                 labelsEmpty,
+			expectedMatch:                       false,
+			expectedSelectorMatchesFunctionCall: false,
+		},
+		"Empty selector with matchEmpty set to true should match everything; attempting to match some labels": {
+			givenSelectorLabels:                 labelsEmpty,
+			matchEmptySelector:                  emptyMatchesEverything(),
+			givenMatchingLabels:                 labelsFiveElements,
+			expectedMatch:                       true,
+			expectedSelectorMatchesFunctionCall: false,
+		},
+		"Empty selector with matchEmpty set to true should match everything; attempting to match empty labels": {
+			givenSelectorLabels:                 labelsEmpty,
+			matchEmptySelector:                  emptyMatchesEverything(),
+			givenMatchingLabels:                 labelsEmpty,
+			expectedMatch:                       true,
+			expectedSelectorMatchesFunctionCall: false,
+		},
+		"More selector labels than received labels -> no match and selector Matches function not called": {
+			givenSelectorLabels:                 labelsThreeElements,
+			matchEmptySelector:                  emptyMatchesEverything(),
+			givenMatchingLabels:                 labelsOneElement,
+			expectedMatch:                       false,
+			expectedSelectorMatchesFunctionCall: false,
```

Comment:
In my test I inject a mockSelector as the labels.Selector of my selectorWrapper. What this mockSelector does is set s.hasMatchesBeenCalled to true if and only if the `internal` Matches function has been called. `selectorWrap.Matches()` calls the Matches function of the selectorWrapper, and in cases where the number of selector labels is greater than the number of item labels the `internal` Matches function is never called. The same goes for all cases where the selectorWrapper contains zero labels, and matchNil is set to false.

Since the entire purpose of this issue is to improve label matching performance by pre-filtering selections by comparing label numbers this test is very important to ensure that this functionality works as expected, in all test cases where its set to false I am confirming that superfluous calls to labels.Selector.Matches() are being avoided.

Reply:
I thought the name was already fine but clearly it is not since it confused you. Perhaps `expectedInternalSelectorMatchesFunctionCall`? Getting very verbose here.